### PR TITLE
Treat md5 usage as errors

### PR DESCRIPTION
**Pull request by sshedi**
This is continuation of PR: https://github.com/twisted/twisted/pull/11883

@adiroiban  please review.

### CODE REVIEW

File: src/twisted/conch/ssh/transport.py [hunk 1]
```diff
@@ -107,11 +101,22 @@ class SSHCiphers:
         b"cast128-ctr": (algorithms.CAST5, 16, modes.CTR),
         b"none": (None, 0, modes.CBC),
     }
+
+    if not fips:
```
Comment (sshedi) [inline src/twisted/conch/ssh/transport.py:105]: this is a major change in this PR.

File: src/twisted/python/compat.py [hunk 2]
```diff
@@ -149,6 +150,29 @@
...
+    Attempt call with 'usedforsecurity=False' if we get a ValueError, which happens when
+    OpenSSL FIPS mode is enabled:
+    ValueError: error:060800A3:digital envelope routines:EVP_DigestInit_ex:disabled for fips
+    """
+
+    try:
+        return hashlib.md5(data, **kwargs)
+    except ValueError as e:
+        if fips:
+            raise Exception("fips is enabled and md5 usage found")
```
Comment (sshedi) [inline src/twisted/python/compat.py:169]: md5 usage is prohibited in fips mode.

**Comment by sshedi** [comment]
I thought adding these changes will be too much in one PR. So made these changes in a new branch.
Strategy is: #11883 should be merged first and then this.

Is this okay?

**Comment by exarkun** [comment]
> I thought adding these changes will be too much in one PR. So made these changes in a new branch. Strategy is: #11883 should be merged first and then this.
> 
> Is this okay?

This PR needs a ticket that describes why something in Twisted needs to change, and possibly what the strategy for making that change it.

**Comment by sshedi** [comment]
> > I thought adding these changes will be too much in one PR. So made these changes in a new branch. Strategy is: #11883 should be merged first and then this.
> > Is this okay?
> 
> This PR needs a ticket that describes why something in Twisted needs to change, and possibly what the strategy for making that change it.

There is one: https://github.com/twisted/twisted/issues/11866

**Comment by exarkun** [comment]
> > > I thought adding these changes will be too much in one PR. So made these changes in a new branch. Strategy is: #11883 should be merged first and then this.
> > > Is this okay?
> > 
> > 
> > This PR needs a ticket that describes why something in Twisted needs to change, and possibly what the strategy for making that change it.
> 
> There is one: #11866

If the two PRs are meant to be merged separately then each one must be a coherent, self-contained unit of changes.  This means each one needs its own ticket describing what that coherent, self-contained unit is meant to address.

**Comment by sshedi** [comment]
> coherent

AFAIK, these two PRs are coherent. The first one allows md5 usage in fips mode and based on some inputs, I decided to treat usage of md5 as error in this PR and modified tests accordingly.

Both PRs are based on same issue solve the same problem.

**Comment by adiroiban** [comment]
Is this an attempt to fix https://github.com/twisted/twisted/issues/11866

I don't understand what is going on in this PR and why so many tests are skipped.

Also, I don't understand why MD5 usage should be treated as an error.

There might be a good reason for all of the above, but please add some notes to the PR description.

At a high level, I understand what you are tying to do here.. but I don't undestand the way you try to fix it.

I have Twisted working with Red Hat OpenSSL FIPS without issues ... but it might be just my particular use case.

As mentioned in the other PR, somehow I don't undestand what is going on here.

Maybe other developers can understand this PR.

If you have time, it might help to hangout over the Matrix/Gitter chat and try to see what can be done to move this forward

https://matrix.to/#/!TlViTjmLDIWAUxzxsO:libera.chat?via=libera.chat&via=gitter.im&via=matrix.org
 
Regards

**Pull request closed without merging**